File: ai-solutions/02/gpt-5.1-codex/python/main.py
import os

def solve():
    # Construct path to input file
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_path = os.path.join(script_dir, '../../../../inputs/02.txt')
    
    with open(input_path, 'r') as f:
        content = f.read().strip()
    
    # Parse ranges
    # Input format: 11-22,95-115,...
    # Can contain newlines based on prompt description, so remove them first or split by comma directly handling whitespace
    ranges = []
    # Replace newlines with nothing just in case, or split by comma and strip
    parts = content.replace('\n', '').split(',')
    
    for p in parts:
        if '-' in p:
            start, end = map(int, p.split('-'))
            ranges.append((start, end))

    total_part1 = 0
    total_part2 = 0

    for start, end in ranges:
        for num in range(start, end + 1):
            s = str(num)
            length = len(s)
            
            # Part 1 check
            is_p1 = False
            if length % 2 == 0:
                half = length // 2
                if s[:half] == s[half:]:
                    is_p1 = True
                    total_part1 += num
            
            # Part 2 check
            # repeated at least twice
            # This means s is composed of a substring p repeating 2 or more times.
            # s = p * k where k >= 2.
            # We iterate possible lengths of p.
            is_p2 = False
            for chunk_len in range(1, length // 2 + 1):
                if length % chunk_len == 0:
                    pattern = s[:chunk_len]
                    repeats = length // chunk_len
                    if pattern * repeats == s:
                        is_p2 = True
                        break
            
            if is_p2:
                total_part2 += num

    print(total_part1)
    print(total_part2)

if __name__ == '__main__':
    solve()

File: ai-solutions/02/gpt-5.1-codex/rust/src/main.rs
use std::fs;
use std::path::Path;

fn main() {
    let paths = ["inputs/02.txt", "../../../../inputs/02.txt"];
    let mut content = String::new();
    let mut found = false;
    
    for p in paths {
        if Path::new(p).exists() {
            content = fs::read_to_string(p).expect("Failed to read file");
            found = true;
            break;
        }
    }
    
    if !found {
        eprintln!("Could not find input file in {:?}", paths);
        std::process::exit(1);
    }

    let mut ranges = Vec::new();
    // Remove newlines and split by comma
    let cleaned_content = content.replace('\n', "");
    for part in cleaned_content.split(',') {
        if part.contains('-') {
            let bounds: Vec<&str> = part.split('-').collect();
            if bounds.len() == 2 {
                if let (Ok(start), Ok(end)) = (bounds[0].parse::<u64>(), bounds[1].parse::<u64>()) {
                    ranges.push((start, end));
                }
            }
        }
    }

    let mut total_part1: u64 = 0;
    let mut total_part2: u64 = 0;

    for (start, end) in ranges {
        for num in start..=end {
            let s = num.to_string();
            let len = s.len();

            // Part 1
            if len % 2 == 0 {
                let half = len / 2;
                if &s[..half] == &s[half..] {
                    total_part1 += num;
                }
            }

            // Part 2
            let mut is_p2 = false;
            for k in 1..=(len / 2) {
                if len % k == 0 {
                    let pattern = &s[..k];
                    let repeats = len / k;
                    // Check if pattern repeated equals s
                    // Efficient way: iterate and check slices
                    // Or construct string
                    if pattern.repeat(repeats) == s {
                        is_p2 = true;
                        break;
                    }
                }
            }

            if is_p2 {
                total_part2 += num;
            }
        }
    }

    println!("{}", total_part1);
    println!("{}", total_part2);
}
